File: 2015/day21.py
"""
Weapons:    Cost  Damage  Armor
Dagger        8     4       0
Shortsword   10     5       0
Warhammer    25     6       0
Longsword    40     7       0
Greataxe     74     8       0

Armor:      Cost  Damage  Armor
Leather      13     0       1
Chainmail    31     0       2
Splintmail   53     0       3
Bandedmail   75     0       4
Platemail   102     0       5

Rings:      Cost  Damage  Armor
Damage +1    25     1       0
Damage +2    50     2       0
Damage +3   100     3       0
Defense +1   20     0       1
Defense +2   40     0       2
Defense +3   80     0       3
"""

from typing import NamedTuple


class Inventory:
    def __init__(self, items):
        self.ring1 = None
        self._armor = None
        self.ring2 = None
        for item in items:
            if isinstance(item, Weapon):
                self.weapon = item
            elif isinstance(item, Armor):
                self._armor = item
            elif isinstance(item, Ring):
                if not self.ring1:
                    self.ring1 = item
                else:
                    self.ring2 = item

    @property
    def names(self):
        return tuple(
            sorted(
                [
                    k.name
                    for k in (self.ring1, self._armor, self.ring2, self.weapon)
                    if k
                ]
            )
        )

    @property
    def not_null(self):
        return [k for k in (self.ring1, self._armor, self.ring2, self.weapon) if k]

    @property
    def damage(self):
        return sum(k.damage for k in self.not_null)

    @property
    def armor(self):
        return sum(k.armor for k in self.not_null)

    def __repr__(self):
        return f"{self.names}"

    def __eq__(self, other):
        return self.names == other.names

    def __hash__(self):
        return hash(self.names)


class Weapon(NamedTuple):
    name: str
    cost: int
    damage: int
    armor: int


class Armor(NamedTuple):
    name: str
    cost: int
    damage: int
    armor: int


class Ring(NamedTuple):
    name: str
    cost: int
    damage: int
    armor: int


rings = [
    Ring(name, int(cost), int(damage), int(armor))
    for name, cost, damage, armor in [
        ("Damage +1,25,1,0".split(",")),
        ("Damage +2,50,2,0".split(",")),
        ("Damage +3,100,3,0".split(",")),
        ("Defense +1,20,0,1".split(",")),
        ("Defense +2,40,0,2".split(",")),
        ("Defense +3,80,0,3".split(",")),
    ]
]

weapons = [
    Weapon(name, int(cost), int(damage), int(armor))
    for name, cost, damage, armor in [
        ("Dagger,8,4,0".split(",")),
        ("Shortsword,10,5,0".split(",")),
        ("Warhammer,25,6,0".split(",")),
        ("Longsword,40,7,0".split(",")),
        ("Greataxe,74,8,0".split(",")),
    ]
]

armors = [
    Armor(name, int(cost), int(damage), int(armor))
    for name, cost, damage, armor in [
        ("Leather,13,0,1".split(",")),
        ("Chainmail,31,0,2".split(",")),
        ("Splintmail,53,0,3".split(",")),
        ("Bandedmail,75,0,4".split(",")),
        ("Platemail,102,0,5".split(",")),
    ]
]


def get_inventories(gold):
    inventories = []
    for weapon in weapons:
        if weapon.cost <= gold:
            rings_gold = gold - weapon.cost
            for ring in rings:
                if ring.cost <= rings_gold:
                    invs_with_ring = get_inventories_ring(
                        rings_gold - ring.cost, [weapon, ring]
                    )
                else:
                    invs_with_ring = []
                invs_without_ring = get_inventories_ring(rings_gold, [weapon])

                inventories.extend(invs_with_ring)
                inventories.extend(invs_without_ring)
    return set([Inventory(inv) for inv in inventories])


def get_inventories_ring(gold, items):
    res = []
    for ring in rings:
        if ring not in items:
            if ring.cost <= gold:
                invs_with_ring = get_inventories_armor(
                    gold - ring.cost, list(items) + [ring]
                )
            else:
                invs_with_ring = []
            invs_without_ring = get_inventories_armor(gold, list(items))

            res.extend(invs_with_ring)
            res.extend(invs_without_ring)
    return res


def get_inventories_armor(gold, items):
    res = [items]
    for armor in armors:
        if armor.cost <= gold:
            res.append(list(items) + [armor])

    return res


"""
Hit Points: 109
Damage: 8
Armor: 2
"""


class Boss(NamedTuple):
    damage: int
    hp: int
    armor: int


PLAYER_HP = 100


def win(inventory, boss):
    eff_boss_damage = max(inventory.damage - boss.armor, 1)
    eff_player_damage = max(boss.damage - inventory.armor, 1)
    boss_hp = boss.hp
    player_hp = PLAYER_HP
    for i in range(boss.hp):
        boss_hp = boss_hp - eff_boss_damage
        if boss_hp <= 0:
            return True
        player_hp = player_hp - eff_player_damage
        if player_hp <= 0:
            return False


gold = 0
boss = Boss(8, 109, 2)
won = False
used = set()
while not won:
    gold += 1
    invs = get_inventories(gold)
    total_invs = len(invs)
    won_invs = 0
    for inventory in invs:
        if inventory not in used:
            used.add(inventory)
        else:
            continue
        if win(inventory, boss):
            won = True
            break

print(gold)

glold = 0
boss = Boss(8, 109, 2)
used = set()
while True:
    gold += 1
    print(gold)
    invs = get_inventories(gold)
    total_invs = len(invs)
    new_invs = 0
    lost_inv = 0
    for inventory in invs:
        if inventory in used:
            continue
        else:
            new_invs += 1
            used.add(inventory)
        if win(inventory, boss):
            continue
        else:
            lost_inv += 1
    if lost_inv == 0 and gold > 111:
        print(f"all wins  with gold {gold} and {total_invs}")
    elif lost_inv > 0 and gold > 111:
        print(f"one lost at {gold}")
    if new_invs == 0 and gold > 111:
        break

File: 2015/day21/day21.go
package main

import (
	"fmt"
)

// Player struct
type Player struct {
	hitPoints int
	damage    int
	armor     int
	inventory []Item
	//items     []Item
}

// Item struct
type Item struct {
	cost   int
	damage int
	armor  int
	name   string
}

var weapons = []Item{
	{name: "Dagger", cost: 8, damage: 4},
	{name: "Shortsword", cost: 10, damage: 5},
	{name: "Warhammer", cost: 25, damage: 6},
	{name: "Longsword", cost: 40, damage: 7},
	{name: "Greataxe", cost: 74, damage: 8},
}

var armors = []Item{
	{name: "Leather", cost: 13, armor: 1},
	{name: "Chainmail", cost: 31, armor: 2},
	{name: "Splintmail", cost: 53, armor: 3},
	{name: "Bandedmail", cost: 75, armor: 4},
	{name: "Platemail", cost: 102, armor: 5},
}

var rings = []Item{
	{name: "Damage+1", cost: 25, armor: 0, damage: 1},
	{name: "Damage+2", cost: 50, armor: 0, damage: 2},
	{name: "Damage+3", cost: 100, armor: 0, damage: 3},
	{name: "Defense+1", cost: 20, armor: 1},
	{name: "Defense+2", cost: 40, armor: 2},
	{name: "Defense+3", cost: 80, armor: 3},
}


func cost(items []Item) int {
	totalCost := 0
	for _, item :=range items {
		totalCost += item.cost
	}
	return totalCost
}
// boss stats
// Hit Points: 109
// Damage: 8
// Armor: 2
// least amount of gold you can spend and still win the fight

func toKey(player Player) string {
	return fmt.Sprintf("%d %d %d", player.damage, player.armor, cost(player.inventory))
}
func main() {
	boss := Player{
		hitPoints: 109,
		damage:    8,
		armor:     2,
	}
	minGold := 8
	maxGold := 1000
	maxGoldToLose := 0
	seen := make(map[string]bool)
	shouldPrint := true
	for gold := minGold; gold < maxGold; gold++ {
		players := getPossiblePlayers(gold)
		for _, player := range players {
			if seen[toKey(player)]{
				continue
			} else {
				seen[toKey(player)] = true
			}
			playerWins := play(player, boss)
			if playerWins && shouldPrint {
				fmt.Printf("with gold = %d the player wins\n%#v\n", gold, player)
				shouldPrint = false
			} else {
				if invCost :=  cost(player.inventory); invCost > maxGoldToLose {
					maxGoldToLose = invCost
				}
			}

		}
	}
	fmt.Println("nMax Gold to lose", maxGoldToLose)
}

func play(player Player, boss Player) bool {
	bossDamage := boss.damage - player.armor
	if bossDamage < 0 {
		bossDamage = 1
	}
	playerDamage := player.damage - boss.armor
	if playerDamage < 0 {
		playerDamage = 1
	}
	playerLife := player.hitPoints
	bossLife := boss.hitPoints
	for {
		bossLife -= playerDamage
		if bossLife <= 0 {
			return true
		}
		playerLife -= bossDamage
		if playerLife <= 0 {
			return false
		}
	}
}

func getPossiblePlayers(gold int) []Player {
	var players []Player
	for _, weapon := range weapons {
		if weapon.cost > gold {
			continue
		}
		// only weapon
		players = append(players, Player{hitPoints: 100, damage: weapon.damage, inventory: []Item{weapon}})

		g := gold - weapon.cost
		// weapon + armor
		for _, armor := range armors {
			if armor.cost > g {
				continue
			}

			// only weapon + armor
			players = append(players, Player{hitPoints: 100, damage: weapon.damage, armor: armor.armor, inventory: []Item{weapon, armor}})
			// weapon + armor + rings
			d := g - armor.cost
			for i := 0; i < len(rings); i++ {
				ring := rings[i]
				if ring.cost > d {
					continue
				}

				// weapon + armor + ring1
				players = append(players, Player{hitPoints: 100, damage: weapon.damage + ring.damage, armor: armor.armor + ring.armor, inventory: []Item{weapon, ring, armor}})
				e := d - ring.cost
				for j := i + 1; j < len(rings); j++ {
					ring2 := rings[j]
					if ring2.cost > e {
						continue
					}
					// weapon + armor + ring + ring2
					players = append(players, Player{hitPoints: 100, damage: weapon.damage + ring.damage + ring2.damage, armor: armor.armor + ring.armor + ring2.armor, inventory: []Item{weapon, ring, ring2, armor}})

				}

			}

		}
		// weapon + rings
		for i := 0; i < len(rings); i++ {
			ring := rings[i]
			if ring.cost > g {
				continue
			}
			// weapon  + ring1
			players = append(players, Player{hitPoints: 100, damage: weapon.damage + ring.damage, armor: ring.armor, inventory: []Item{weapon, ring}})
			e := g - ring.cost
			for j := i + 1; j < len(rings); j++ {
				ring2 := rings[j]
				if ring2.cost > e {
					continue
				}
				// weapon + armor + ring + ring2
				players = append(players, Player{hitPoints: 100, damage: weapon.damage + ring.damage + ring2.damage, armor: ring.armor + ring2.armor, inventory: []Item{weapon, ring, ring2}})

			}

		}
	}

	return players
}
